### ISSUE TITLE
"Cannot register Static Global Var" Error when loading HIP code

### ISSUE THREAD

**Issue by fweig** [ROCm-Developer-Tools/HIP#3201]
I'm trying load a shared library containing HIP code at runtime via `dlopen`. 

Doing so crashes the application with a "Cannot register Static Global Var" Error from the HIP runtime.

Are the any pointers what could cause this / how to solve this?

**Comment by b-sumner** [comment]
Is your shared library linked to the hip runtime, i.e. are all of its dependencies explicitly linked in?

**Comment by fweig** [comment]
I believe so? `ldd` gives:
```
        linux-vdso.so.1 (0x00007ffc2b9cc000)
        libamdhip64.so.5 => /opt/rocm-5.4.3/hip/lib/libamdhip64.so.5 (0x00007f957eaf1000)
        libstdc++.so.6 => /lib/x86_64-linux-gnu/libstdc++.so.6 (0x00007f957e8ad000)
        libm.so.6 => /lib/x86_64-linux-gnu/libm.so.6 (0x00007f957e7c6000)
        libgcc_s.so.1 => /lib/x86_64-linux-gnu/libgcc_s.so.1 (0x00007f957e7a6000)
        libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007f957e57c000)
        libhsa-runtime64.so.1 => /opt/rocm-5.4.3/lib/libhsa-runtime64.so.1 (0x00007f957e2d3000)
        libnuma.so.1 => /lib/x86_64-linux-gnu/libnuma.so.1 (0x00007f957e2c6000)
        /lib64/ld-linux-x86-64.so.2 (0x00007f9580482000)
        libelf.so.1 => /lib/x86_64-linux-gnu/libelf.so.1 (0x00007f957e2a8000)
        libdrm.so.2 => /lib/x86_64-linux-gnu/libdrm.so.2 (0x00007f957e290000)
        libdrm_amdgpu.so.1 => /lib/x86_64-linux-gnu/libdrm_amdgpu.so.1 (0x00007f957e284000)
        libz.so.1 => /lib/x86_64-linux-gnu/libz.so.1 (0x00007f957e268000)
```

In CMake it's just linked against `hip::device` and `hip::hipcub`.

**Comment by b-sumner** [comment]
But is it NEEDED?  "readelf -d your_shared.so | grep NEEDED"

**Comment by fweig** [comment]
With your command, i get:
```
 0x0000000000000001 (NEEDED)             Shared library: [libamdhip64.so.5]
 0x0000000000000001 (NEEDED)             Shared library: [libstdc++.so.6]
 0x0000000000000001 (NEEDED)             Shared library: [libm.so.6]
 0x0000000000000001 (NEEDED)             Shared library: [libgcc_s.so.1]
 0x0000000000000001 (NEEDED)             Shared library: [libc.so.6]
```

**Comment by b-sumner** [comment]
Thanks!  I guess we need to dig deeper.  Can you provide a minimal reproducer?

**Comment by fweig** [comment]
Writing a reproducer turned out to be much easier than I expected. Here it is:
```c++
#include <hip/hip_runtime.h>
#include <iostream>

#define HIP_CALL(x) do { \
    hipError_t err = x; \
    if (err != hipSuccess) { \
        std::cerr << "Error: " << hipGetErrorString(err) << std::endl; \
        exit(1); \
    } \
} while (0)

struct Foo {
    int x;
    __host__ __device__ Foo() {}
    __host__ __device__ ~Foo() {}
};

template<typename T>
__constant__ T cmem;

__global__ void someKernel(int *x) {
    *x = cmem<Foo>.x;
}

int main() {
    Foo foo;
    foo.x = 42;
    HIP_CALL(hipMemcpyToSymbol(HIP_SYMBOL(cmem<Foo>), &foo, sizeof(Foo)));

    int *dx;
    HIP_CALL(hipMalloc(&dx, sizeof(int)));
    someKernel<<<1, 1>>>(dx);

    int x;
    HIP_CALL(hipMemcpy(&x, dx, sizeof(int), hipMemcpyDeviceToHost));

    if (x == 42) {
        std::cout << "SUCCESS!" << std::endl;
    } else {
        std::cout << "FAILURE!" << std::endl;
    }

    HIP_CALL(hipFree(dx));
    return 0;
}
```

Expected outcome: Application prints `SUCCESS!`
Actual outcome: Applications crashes with aforementioned `"Cannot register Static Global Var"` error

Compiled with AMD Clang (ROCm 5.4.3).

Fwiw, there are two ways to circumvent the issue:

1. Marking the ctor/dtor of `Foo` as `default` instead:
```c++
    __host__ __device__ Foo() = default;
    __host__ __device__ ~Foo() = default;
```

2. Or removing the template from `cmem`: 
```
__constant__ Foo cmem;
```

In both cases the application runs without issue.

**Comment by b-sumner** [comment]
Thank you!  We'll take a look.

**Comment by ppanchad-amd** [comment]
@b-sumner Has this issue been resolved? Thanks!

**Comment by tcgu-amd** [comment]
Hi @fweig, thank you for reaching out., and sorry for the lack of responses. The issue you are experiencing is likely due to adding a `__constant__` qualifier to a template. Like the error message suggests, you can look at the `__constant__` qualifier as essentially adding a static global variable. In general, defining a static global variable template is a bad idea and is illegal in regular c++ due to the ambiguity it introduces. Although how `__constant__` works is different from plain `static global`, the same principle applies since it is trying to achieve the same effect across both host and device memories. 

To understand why, we must know that HIP uses the [registerStatGlobalVar](https://github.com/ROCm/clr/blob/ade195401580498f0726520827c3fd0a8ae265d9/hipamd/src/hip_code_object.cpp#L1257) function to register a shadow variable on the host for any `__constant__`  variable instances of non-Plain-Old-Datatype (POD as per C++11 standard) in order to keep track of the variable across kernel instances.

Normally, this is quite straight forward and happens at variable instantiation (e.g. at declaration `__constant__ Foo foo;`), but in your case there is no instance at `__constant__ T cmem` due to the template. Instead, the template is first instantiated in your kernel code. If the kernel were a regular host function, the variable would be instantiated there as a global variable and the main host function would be able to refer to it, so there won't be a problem. However, a kernel function does not share the same variable scope as the host, which means that `cmem<Foo>` would be instantiated twice as separate variables, once in kernel, and once in host code.

The problem with this is that [registerStatGlobalVar](https://github.com/ROCm/clr/blob/ade195401580498f0726520827c3fd0a8ae265d9/hipamd/src/hip_code_object.cpp#L1257) ends up being called twice as well, on two variables with the same symbol `cmem<Foo>`.
I ran your reproducer quickly and this is indeed the case, `cmem<foo>` is being registered twice which causes an error on the second call.

As to why your fixes/workarounds work, it is also quite straightforward. By using `default` as the ctor/dtor, you reduce `Foo` to POD, which does not require calling registerStatGlobalVar at all. In the other case, by removing the template, you avoid the issue of static global templating, which also fixes the problem.

tldr; *don't use `__constant__` with templates*. The code is behaving as expected; doing something like this in C++ in general does not make sense. 

Anyways, this issue is going to be closed, but please feel free to ask follow up questions. Thanks!!
